apps/passport-server/src/services/generic-issuance/pipelines/LemonadePipeline.ts

```diff
@@ -1941,6 +1969,377 @@ export class LemonadePipeline implements BasePipeline {
+    for (const checkerTicketAtom of await this.db.loadByEmail(
+      this.id,
+      checkerEmail
+    )) {
+      checkerProductIds.push(checkerTicketAtom.productId);
+    }
+    for (const manualTicket of this.getManualTicketsForEmail(checkerEmail)) {
+      checkerProductIds.push(manualTicket.productId);
+    }
+
+    // Return any event/product combinations for which the user has a superuser
+    // ticket.
+    return this.definition.options.events.reduce((memo, eventConfig) => {
+      if (
+        // Does the user own a superuser product for this event?
+        eventConfig.ticketTypes.some(
+          (ticketType) =>
+            ticketType.isSuperUser &&
+            checkerProductIds.includes(ticketType.genericIssuanceProductId)
+        )
+      ) {
+        // In that case, they can check in any of the product types for this event
+        memo.push(
+          ...eventConfig.ticketTypes.map((ticketType) => ({
+            eventId: eventConfig.genericIssuanceEventId,
+            productId: ticketType.genericIssuanceProductId
+          }))
+        );
+      }
+      return memo;
+    }, [] as MemberCriteria[]);
+  }
+
+  /**
+   * Returns offline tickets that the user is entitled to check in, if any.
+   */
+  public async getOfflineTickets(
+    checkerEmail: string
+  ): Promise<PodboxOfflineTicket[]> {
+    return traced(LOG_NAME, "getOfflineTickets", async (span) => {
+      tracePipeline(this.definition);
+
+      if (this.definition.options.offlineCheckin === undefined) {
+        throw new PCDHTTPError(
+          401,
+          "Offline check-in is not enabled for this pipeline"
+        );
+      }
+
+      const checkerTicketCriteria =
+        await this.ticketsUserCanCheckIn(checkerEmail);
+      if (checkerTicketCriteria.length === 0) {
+        span?.setAttribute("offline_tickets_returned", 0);
+        // User can't check anything in, so they don't get any offline tickets.
+        return [];
+      }
+
+      const enabledFields = this.definition.options.offlineCheckin.fields;
+      const offlineTickets: PodboxOfflineTicket[] = [];
+
+      for (const atom of await this.db.load(this.id)) {
+        if (!atom.email) {
+          continue;
+        }
+
+        // If the user can check in tickets of this type
+        if (ticketMatchesCriteria(atom, checkerTicketCriteria)) {
+          offlineTickets.push({
+            id: atom.id,
+            eventId: atom.eventId,
+            eventName: this.lemonadeAtomToEventName(atom),
+            ticketName: this.lemonadeAtomToTicketName(atom),
+            attendeeEmail: enabledFields.attendeeEmail
+              ? atom.email.toLowerCase()
+              : undefined,
+            attendeeName: enabledFields.attendeeName ? atom.name : undefined,
+            checker: enabledFields.checker ? null : undefined,
+            is_consumed: atom.checkinDate instanceof Date
+          });
+        }
+      }
+
+      if ((this.definition.options.manualTickets ?? []).length) {
```

Comment:
we'll need to be sure to not forget to make this work with the database-backed manual tickets once they get implemented for `LemonadePipeline`

Reply:
Yes. At some point we might want to stop adding features that could/should be generic to just one pipeline, because then when we *do* add generic features like offline check-in, we have to code around all of the differences between the pipeline types due to them having different features. Ideally the only difference between a Pretix and Lemonade pipeline would be which back-end they get their ticket data from.